Removed config maps and replaced with rados omaps

ShyamsundarR opened the pull request:
Existing config maps are now replaced with rados omaps that help
store information regarding the requested volume names and the rbd
image names backing the same.

Further to detect cluster, pool and which image a volume ID refers
to, changes to volume ID encoding has been done as per provided
design specification in the stateless ceph-csi proposal.

ShyamsundarR commented:
Testing done includes,

    Create delete PVC
    Create pods using PVC
    Create/promote/use/delete snapshots
    Create/delete 5 PVCs

TODO items are as follows:

    Update documentation across the repository
    Changes to deployment files
    Changes to examples directory

Tech Debt left for later (i.e optimizations and cleanup that does not hamper functionality and are not going to be a part of this PR):

    Create a cache for pool IDs, instead of fetching them each time
    Create a cache for secrets and move into watch mode for the same
    Unit testing cephcmds.go
    Fetch omaps in a list than one by one
    Provide tools to inspect omaps against the Ceph cluster
    Design and implement List[Volumes|Snapshots] verbs, including deciding on using the rbdImageOMap to store created snapshots, in order to support deletion of source images even when they have snapshots

ShyamsundarR commented:
Moved the PR against master from the older one against v1.0 branch.

@gman0 @dillaman @Madhu-1 PTAL

ShyamsundarR commented:
@gman0 @Madhu-1 @dillaman PTAL, added one more commit to change documentation and sample manifests. Pending change now would be Helm charts (and catching upto master).

There is an urgency to try and get this in, so that we can move toward a 1.0 release and also adopt the same into the Rook projects 1.0 release (in 2 weeks as of this writing). Request attention with this consideration in mind. Thanks!

dillaman commented:
@ShyamsundarR Ack -- I'll start reviewing this tonight.

dillaman commented on docs/deploy-rbd.md:
```diff
@@ -32,51 +32,45 @@ Option | Default value | Description
-`csi.storage.k8s.io/provisioner-secret-name`, `csi.storage.k8s.io/node-publish-secret-name` | for Kubernetes | name of the Kubernetes Secret object containing Ceph client credentials. Both parameters should have the same value
-`csi.storage.k8s.io/provisioner-secret-namespace`, `csi.storage.k8s.io/node-publish-secret-namespace` | for Kubernetes | namespaces of the above Secret objects
+`csi.storage.k8s.io/provisioner-secret-name`, `csi.storage.k8s.io/node-publish-secret-name` | (no) for Kubernetes | name of the Kubernetes Secret object containing Ceph client credentials. Both parameters should have the same value
+`csi.storage.k8s.io/provisioner-secret-namespace`, `csi.storage.k8s.io/node-publish-secret-namespace` | (no) for Kubernetes | namespaces of the above Secret objects
```
Would the admin ID/key ever be stored in a user's namespace? The example shows both the admin and user credentials in the same yaml whereas the in-tree driver only required the user credentials in the user's namespace.

dillaman replied:
... I am also confused again since looking at ```K8sConfig```, it is hard-coded to pull secrets named ```ceph-cluster-<cluster-id>```. The documentation for ```external-provisioner``` [1] shows that the secrets specified here will be passed in ```CreateVolumeRequest.secrets```/```DeleteVolumeRequest.secrets```/```NodePublishVolumeRequest.secrets``` -- so to me it's confusing that some places are pulling secrets from ```ceph-cluster-<cluster-id>``` and others are getting overrides from the ```external-provisioner```?

[1] https://kubernetes-csi.github.io/docs/print.html#createdelete-volume-secret

ShyamsundarR replied:
Resolved as per [comments](https://github.com/ceph/ceph-csi/pull/312#issuecomment-482631799) raised on the PR around the solution.

dillaman commented on pkg/rbd/rbd_util.go:
```diff
@@ -192,13 +167,8 @@ func rbdStatus(pOpts *rbdVolume, userID string, credentials map[string]string) (
-	klog.V(4).Infof("rbd: status %s using mon %s, pool %s", image, mon, pOpts.Pool)
-	args := []string{"status", image, "--pool", pOpts.Pool, "-m", mon, "--id", userID, "--key=" + key}
+	klog.V(4).Infof("rbd: status %s using mon %s, pool %s", image, pOpts.Monitors, pOpts.Pool)
+	args := []string{"status", image, "--pool", pOpts.Pool, "-m", pOpts.Monitors, "--id", userID, "--key=" + key}
```
I know you didn't introduce the usage of ```--key``` to all the CLI tools, but is there any concern that the keys are visible to anyone running a "ps"? It would seem like writing the key to a locked-down temp file and using the ```--keyfile``` argument would be a better approach.

ShyamsundarR replied:
Yes this is a concern, @phlogistonjohn had pointed this out as well in a separate conversation. This needs to be handled in a subsequent PR, for which a tracking issue does not exist yet (I will create one later today, if one of you hasn't by then).

dillaman replied:
Thanks, you validated my concern -- opened #318

dillaman commented on examples/rbd/storageclass.yaml:
```diff
@@ -34,20 +25,19 @@ parameters:
-   # NOTE: If using "clusterID" instead of "monitors" above, the following
-   # IDs MAY be added to the ceph-cluster-<cluster-id> secret and skipped
-   # here
+   # NOTE: If the secrets above are skipped, then the following values are
```
Shouldn't it pull the adminid and userid from the secret? Why are there multiple places to store this data?

ShyamsundarR replied:
Resolved the text, this was left over comments when the secret was removed.

dillaman commented on pkg/rbd/rbd_util.go:
```diff
@@ -39,35 +45,45 @@ const (
+	ImageFormat        string
+	ImageFeatures      string
+	VolSize            int64
+	AdminID            string
```
Shouldn't ```AdminID```,```UserID```,```Monitors``` always be pulled from the secrets?

ShyamsundarR commented:
Secrets/keys/IDs discussion

So to begin with, some of the questions you raise are concerns to me as well, and (partly?) covered in this [issue](https://github.com/ceph/ceph-csi/issues/270). I am summarizing various comments around the topic here to provide a current, and possibly discuss a (near) future, answer to the same.

I would start referring to the 2 secrets (they are both kubernetes secrets, yes) as the secret and the config, where the secret is the one mentioned in the StorageClass [1] and config is the Ceph cluster config (ceph-cluster-<cluster-id>) provided to the CSI plugin. Keeps the further discussion less ambiguous referring to the terms.

I am also throwing in the term credential in the discussion to refer to an ID/key (as in Ceph user ID and key) in the discussion below.

1) The [issue](https://github.com/ceph/ceph-csi/pull/312#discussion_r274727544) of the ID being separate from the key in the StorageClass and in the secret.yaml, is a usability issue IMO, and this should be fixed (as a part of addressing #270)

2) Would the credentials in the secret be stored in the user's namespace?

No. The secret names and namespaces that are passed down from the StorageClass need to exist in **a** namespace, that is read by the node level services that invoke the CSI plugin RPCs with the data from these secrets. Hence, user namespaces need not have access to these secrets.

Also, the config secret needs to be readable by the CSI plugin on the nodes, and hence needs appropriate access to the secret in whichever namespace they exist.

In both cases, the user namespace requesting the volume provisioning or mounting service, do not need access to the secrets, nor are these present in their namespace.

3) Why credentials in 2 places (secrets and config)?

If we want to support fine grained credentials per pool, I.e credentials that have access to do CRUD operations on a pool and not across pools in use by the CSI plugin, then we would want to use the credentials from the secrets in the StorageClass, else we can use the credentials in the config itself.

For list operations, we will not get the credentials from the secrets and have to rely on the credentials from the config solely. If the config credentials can be crafted to have list permissions (IOW metadata read) then we can use the same for listing, else listing would have to remain unsupported.

4) Why 2 credentials, the admin and the user/mounter?

Again, if we can segregate permissions for these credentials, with the user credential having only mount permissions (post mount reads and writes to the image should be possible) and the admin credential having CRUD permissions, it makes sense to have 2, else we should just fold these into a single credential. (part of #270)

5) Where are credentials pulled from?

The order of using credentials in various RPCs are as follows,
- Use credentials passed in by the RPC if they exist
- Fall back to using credentials from the config if the RPC does not provide the credentials
- Error, if both locations to not have credentials

I think we need to continue this in #270 and look at the need for 2 credentials (user/admin) and then see if we want per-pool level credential segregation and if so retain the current scheme, else throw away the credentials from being passed in as secrets.

[1] https://kubernetes-csi.github.io/docs/print.html#createdelete-volume-secret

dillaman commented:
> I would start referring to the 2 secrets (they are both kubernetes secrets, yes) as the secret and the config, where the secret is the one mentioned in the StorageClass [1] and config is the Ceph cluster config (ceph-cluster-) provided to the CSI plugin. Keeps the further discussion less ambiguous referring to the terms.

My concern is that they *both* provide credentials and that is confusing. I don't think there should be two different places to store the keys and up to three places to store the user ID. The fact that the code sometimes pulls the keys from the k8sconfig store via the ```cluster-id``` means it's switching back-and-forth between credentials potentially.

> 2. Would the credentials in the secret be stored in the user's namespace?
>
> No. 

Ack -- cool, thx. 

> 1. Why credentials in 2 places (secrets and config)?
> 
> If we want to support fine grained credentials per pool, I.e credentials that have access to do CRUD operations on a pool and not across pools in use by the CSI plugin, then we would want to use the credentials from the secrets in the StorageClass, else we can use the credentials in the config itself.

Sorry, I wasn't really asking about the difference between "admin" and "user" -- just the fact that the "admin"/"user" secrets are stored twice (and the admin/user IDs are also specified multiple times).
 
> 1. Where are credentials pulled from?
> 
> The order of using credentials in various RPCs are as follows,
> 
> * Use credentials passed in by the RPC if they exist
> * Fall back to using credentials from the config if the RPC does not provide the credentials
> * Error, if both locations to not have credentials

... and then fallback to the ID overrides from the StorageClass if not specified in the config secret? 
 
> I think we need to continue this in #270 and look at the need for 2 credentials (user/admin) and then see if we want per-pool level credential segregation and if so retain the current scheme, else throw away the credentials from being passed in as secrets.

Thanks for the pointer to that issue. I agree the difference between admin/user credentials is a separate issue. I am just worried about how confusing it is w/ all the different places to store a given set of credentials and how some CSI operations would pull credentials from the secret vs others that would pull it from the config.

ShyamsundarR commented:
> Thanks for the pointer to that issue. I agree the difference between admin/user credentials is a separate issue. I am just worried about how confusing it is w/ all the different places to store a given set of credentials and how some CSI operations would pull credentials from the secret vs others that would pull it from the config.

Discussed this further with @dillaman and here is the gist of the same:
- We remove credentials from the config (secret) store `ceph-cluster-<clusterid>`
- We will not support List operations,
  - These are currently CSI driver generated operations, and do not impact CO systems
  - Further are optional from a CSI support perspective as well
- The config store will also remove `pools` as this was present for listing
- Thus the config store will no longer be a secret as it contains only monitors, and this is public information that need not be stored as a secret
- The config store will also move to a single config map as in [solution alternative 1](https://github.com/ceph/ceph-csi/blob/e6139baaff4eca8209b38c514d37ac49b544a567/docs/design/proposals/stateless-csi-plugin.md#solution-alternative-1) as we are not storing secrets and hence the con regarding the secret management will not happen.
  - The advantage of the same is that we need not implement any kubernetes watchers and such in code to keep the in-memory config store upto date with changed monitor information
- The StorageClass will continue to mention `ClusterID`, and this ID would be found in the config map, with required monitor information in it

dillaman commented:
@ShyamsundarR I was digging through the CSI spec and noticed the ```EXPAND_VOLUME``` cap. That seems like something that the RBD CSI should support (the in-tree driver supports resizing). However, the request seems to only include the ```VolumeID``` and no additional parameters/secrets to connect to the storage system. The workaround might be to eventually require ```STAGE_UNSTAGE_VOLUME``` on the node controller and cache the credentials in a temp file for use by the expand request.

The ```GET_CAPACITY``` and ```GET_VOLUME_STATS``` seem like another odd case where we won't get any secrets. The capacity RPC at least includes the StorageClass parameters so it could make a request of the secrets perhaps. The volume stats RPC gives a path to a published volume, so like the expand volume RPC, perhaps the credentials would be cached to a local file?

ShyamsundarR commented:
1) The RPC ControllerExpandVolume supports secrets afaict [1](https://github.com/container-storage-interface/spec/blob/master/lib/go/csi/csi.pb.go#L3377) [2](https://kubernetes-csi.github.io/docs/secrets-and-credentials.html#csi-operation-secrets), are you talking about NodeExpandVolume? If so,
What would we need to do on the node to expand the volume? (this is if we support online resizing, which I think we do)
- The rbd image is already mapped
- If it is not a Block volume, there is an FS mounted on the image

I was assuming the image expansion is auto-detected on the rbd clients, and the FS may need growfs like operations. Thus, we really do not need credentials on the NodeExpandVolume. Am I missing something here?

2) NodeGetVolumeStats is fetching stats for already mounted volumes, and the returned data seems to want, inodes OR bytes used/available/total per mounted volume. For which it can use the mounted image to get the required data, right? It again does not need credentials to fetch data from the Ceph cluster I would assume.

3) GetCapacity stands unimplemented as of now (and is optional as noted). Also, the passed in `parameters` field (i.e storage class details) are further optional in the spec. so there is no real manner in which it can be determined which entities capacity is required afaict. I would leave this unimplemented for the future as well.

dillaman commented:
Re: (1) I think I was just looking at the NodeExpandVolumeRequest and didn't see the ControllerExpandVolumeRequest. Disregard.

Re: (2) I know there is a PR to get "metrics", but it looks like it's only being added to the CephFS driver for NodeGetVolumeStats. If there isn't any need to know the provisioned vs actual byte specs for RBD block PVs, I guess it's not needed.

Re: (3) If it's not used by k8s, no worries. I am just trying to think ahead for worst-case.

ShyamsundarR commented:
> Re: (1) I think I was just looking at the NodeExpandVolumeRequest and didn't see the ControllerExpandVolumeRequest. Disregard.
> 
> Re: (2) I know there is a PR to get "metrics", but it looks like it's only being added to the CephFS driver for NodeGetVolumeStats. If there isn't any need to know the provisioned vs actual byte specs for RBD block PVs, I guess it's not needed.

For block, I would assume we get metrics from the filesystem (or consumer) on the block device. Does it make sense to return used/free bytes from a block device?

> 
> Re: (3) If it's not used by k8s, no worries. I am just trying to think ahead for worst-case.

This is useful, so thanks. I had done this exercise earlier (i.e looking at secrets and which RPCs etc.) but helps with another review.

kfox1111 commented:
I do really like the idea of breaking the somewhat strange loop in the driver that calls back to k8s. Big thank you for working on a fix. :)

I didn't see it called out up above, but one critical piece will be a migration path for those of us already using it. Is this planned?

batrick commented on pkg/rbd/rbd.go:
```diff
@@ -27,10 +27,98 @@ import (
+  - stores keys named using the CO generated names for volume requests
+  - keys are named "csi.volume."+[CO generated VolName]
+  - Key value contains the RBD image uuid that is created or will be created, for the CO provided
+  name
```
Whenever I see applications making raw use of omaps I get nervous. The reason for this is mostly around the de facto limit of ~100,000 k/v pairs per object. Are we confident that a cluster won't have more than 100k volumes?

ShyamsundarR replied:
Fair enough :) The initial numbers I discussed with @dillaman were 100K to be exact.

Anyway, as this information is new to me, and we added to omap to support listing (i.e bunch up the csi volumes under an omap to help with listing). My initial solution to the problem (as we can never answer what scale this may reach) would be to remove the `csi.volumes.` and `csi.snaps.` omap, but retain the volname and volid based omaps and keys, as we are no longer intending to support listing RPCs. Thoughts?

(the other knee jerk reaction was, let's create chapters of csi.volumes and use the same, but that depends on if we do not have to maintain counters around number of keys in an omap and can work with errors returned from a omap setomapval instead, that can help understand it is "full")

ShyamsundarR replied:
Currently we can scale upto 100K, which is not essentially a hard ceiling. IOW, the updates to the objects may slow down, but not fail at that exact number.


This was discussed, and a possible path for the future is noted in [this](https://github.com/ceph/ceph-csi/pull/312#issuecomment-487949077) comment.

IOW, we decided to continue with the current scale, and in the future if required shard the objects, which is feasible under the current scheme.

batrick commented on pkg/rbd/rbd.go:
```diff
@@ -27,10 +27,98 @@ import (
+  - stores keys named using the CO generated names for snapshot requests
+  - keys are named "csi.snap."+[CO generated SnapName]
+  - Key value contains the RBD snapshot uuid that is created or will be created, for the CO
+  provided name
```
Same concern: k/v pair per snapshot won't scale I think.

batrick commented on examples/README.md:
```diff
@@ -1,22 +1,60 @@
+  * Key is typically the output of, `ceph auth get-key client.admin` where
+    `admin` is the client Admin ID
+  * This is used to substitute `[admin|user]id` and `[admin|user]key` values in
+    the template file
```
I'm not seeing a particular reason why the ceph-csi needs the admin key. Is it so that it can do `ceph osd lspools`?

ShyamsundarR replied:
The reason was a simple carryover of what existed prior to this PR. The exact credentials and its permissions are being worked out in issue #270 when we also need to think about what rights are required for the various operations being performed (like `ceph osd lspools`).

ShyamsundarR commented:
> > Thanks for the pointer to that issue. I agree the difference between admin/user credentials is a separate issue. I am just worried about how confusing it is w/ all the different places to store a given set of credentials and how some CSI operations would pull credentials from the secret vs others that would pull it from the config.
> 
> Discussed this further with @dillaman and here is the gist of the same:
> 
>     * We remove credentials from the config (secret) store `ceph-cluster-<clusterid>`
> 
>     * We will not support List operations,
>       
>       * These are currently CSI driver generated operations, and do not impact CO systems
>       * Further are optional from a CSI support perspective as well
> 
>     * The config store will also remove `pools` as this was present for listing
> 
>     * Thus the config store will no longer be a secret as it contains only monitors, and this is public information that need not be stored as a secret
> 
>     * The config store will also move to a single config map as in [solution alternative 1](https://github.com/ceph/ceph-csi/blob/e6139baaff4eca8209b38c514d37ac49b544a567/docs/design/proposals/stateless-csi-plugin.md#solution-alternative-1) as we are not storing secrets and hence the con regarding the secret management will not happen.
>       
>       * The advantage of the same is that we need not implement any kubernetes watchers and such in code to keep the in-memory config store upto date with changed monitor information
> 
>     * The StorageClass will continue to mention `ClusterID`, and this ID would be found in the config map, with required monitor information in it

@dillaman Addressed the dual credential issue with the latest commit, PTAL. The issue of still having to specify 2 credentials (user and admin) and also the ID separate from the Key would be dealt with as a part of fixing #270 

@gman0 @Madhu-1 Configuration requirements have changed from requiring a series of values in a [secret](https://github.com/ceph/ceph-csi/blob/13c60eae7cfdf853172f267c7b80a8816ac59577/examples/rbd/template-ceph-cluster-ID-secret.yaml#L1) to a config map of cluster IDs containing the [monitors](https://github.com/ceph/ceph-csi/pull/312/files?file-filters%5B%5D=.go&file-filters%5B%5D=.yaml#diff-77ed618807e98a36e2f5c1ac9cc9c608).

ShyamsundarR commented:
> I do really like the idea of breaking the somewhat strange loop in the driver that calls back to k8s. Big thank you for working on a fix. :)
> 
> I didn't see it called out up above, but one critical piece will be a migration path for those of us already using it. Is this planned?

Yes, migration is not planned as yet. This has cropped up from a 0.3 version to this instance as well in #296 

Should we discuss the same in the other issue to arrive at a more generic solution? Also, at present as migration solutions I do not have any thoughts to offer.

dillaman commented:
> Yes, migration is not planned as yet. This has cropped up from a 0.3 version to this instance as well in #296
> 
> Should we discuss the same in the other issue to arrive at a more generic solution? Also, at present as migration solutions I do not have any thoughts to offer.

It would be great if a generic migration solution would fit here assuming the v0.3 driver used a different storage class. That same solution could then be used for migrating volumes between different storage tiers (i.e. bronze, silver, gold) or between non-mirrored and mirrored.

ShyamsundarR commented:
Updated the Helm charts and squashed all commits into one.

Among the open issues, is only deciding about the omap scale as per this [comment](https://github.com/ceph/ceph-csi/pull/312#discussion_r275506626).

I am hoping we can get this in this week, and start work on credential changes to RBD and also working on the CephFS plugin to incorporate similar changes.

Requesting reviews, @dillaman @Madhu-1 @gman0 (also, @batrick from the omap usage perspective (among others)).

kfox1111 commented:
Kind of concerned.... what is the migration path here? There are existing users of the chart.

ShyamsundarR commented:
> Among the open issues, is only deciding about the omap scale as per this [comment](https://github.com/ceph/ceph-csi/pull/312#discussion_r275506626).

On further discussions regarding the omap scale issue, the resolution would be to apply sharding (as suggested by @liewegas) to the volumes and snaps omaps **when required**.

The ability to shard can be based on selecting a shard using the `volumename` and the `snapname` hash, to map to a shard number, where the shards are omaps with the specified shard number in their name.

Also, when sharding is introduced, the existing omaps may need a one-time "split into shards" operation to be performed, but this **will not** impact the metadata exchanged with the CO environment.

Added sharding to the [TechDebt list](https://github.com/ceph/ceph-csi/pull/312#issuecomment-480793273) as well.

dillaman commented on pkg/rbd/rbd_util.go:
```diff
@@ -205,10 +185,491 @@ func rbdStatus(pOpts *rbdVolume, userID string, credentials map[string]string) (
+// reserveOMapName creates an omap with passed in oMapNameSuffix and a generated <uuid>.
+// It ensures generated omap name does not already exist and if conflicts are detected, a set
+// number of retires with newer uuids are attempted before returning an error
+func reserveOMapName(monitors, adminID, key, poolName, oMapNameSuffix string) (string, error) {
```
Nit: isn't this a prefix instead of a suffix?

ShyamsundarR replied:
Fixed.

dillaman commented on pkg/rbd/rbd_util.go:
```diff
@@ -467,141 +1034,252 @@ func protectSnapshot(pOpts *rbdSnapshot, adminID string, credentials map[string]
+	ID        int64  `json:"id"`
+	Name      string `json:"name"`
+	Size      int64  `json:"size"`
+	TimeStamp string `json:"timestamp"`
```
Nit: Timestamp is one word

ShyamsundarR replied:
Fixed.

ShyamsundarR commented on examples/rbd/csi-config-map-sample.yaml:
```diff
@@ -0,0 +1,29 @@
+# identified by the <cluster-id>
+# If a CSI plugin is using more than one Ceph cluster, repeat the "<cluster-id>"
+# section for each such cluster in use.
+# NOTE: If MON addresses change, this config map can be updated and replaced
```
Short answer: No, a restart/redeploy of the pods is not required

The config map contents would be updated within the container by Kubernetes, on a periodic basis. Code always opens and reads the config file, and thus would get refreshed contents when Kubernetes updates the same.

Also, I think I will add instructions on how to achieve this in Kubernetes in the README.md for reference (and maybe in the sample config map as well).

(other comments would be addressed in the next commit)

ShyamsundarR replied:
Fixed. Updated comments appropriately, also README.md felt sufficient and was not modified for this.

ShyamsundarR commented on docs/deploy-rbd.md:
```diff
@@ -32,52 +32,39 @@ Option | Default value | Description
-output of `ceph fsid` of the Ceph cluster to be used for provisioning.
+**NOTE:** An accompanying CSI configuration file, needs to be provided to the
+running pods. Refer to [Creating CSI configuration for RBD based provisioning]
+(../examples/README.md#creating-csi-configuration-for-rbd-based-provisioning)
```
Fixed. Line split between text and link caused the issue.

ShyamsundarR commented on docs/deploy-rbd.md:
```diff
@@ -101,6 +88,18 @@ Those manifests deploy service accounts, cluster roles and cluster role
+The config map deploys an empty CSI configuration that is mounted as a volume
+within the Ceph CSI plugin pods. To add a specific Ceph clusters configuration
+details, refer to [Creating CSI configuration for RBD based provisioning]
+(../examples/README.md#creating-csi-configuration-for-rbd-based-provisioning)
```
Fixed. Line split between text and link caused the issue.

ShyamsundarR commented on examples/rbd/csi-config-map-sample.yaml:
```diff
@@ -0,0 +1,29 @@
+# The <cluster-id> is used by the CSI plugin to uniquely identify and use a
+# Ceph cluster, the value MUST match the value provided as `clusterID` in the
+# StorageClass
+# The <MONValuei> fields are the various monitor addresses for the Ceph cluster
```
Fixed. Trailing "i" was intended to denote the number, added a "#" instead.

ShyamsundarR commented on examples/rbd/snapshotclass.yaml:
```diff
@@ -5,19 +5,16 @@ metadata:
   # cannot be greater than 36 bytes in length, and should remain immutable for
   # the lifetime of the StorageClass in use.
-  # If using clusterID, ensure to create a secret, as in
+  # Ensure to create a secret named ceph-cluster-<cluster-id>, based on
```
Fixed.

ShyamsundarR commented on examples/rbd/storageclass.yaml:
```diff
@@ -5,23 +5,14 @@ metadata:
    # cannot be greater than 36 bytes in length, and should remain immutable for
    # the lifetime of the StorageClass in use.
-   # If using clusterID, ensure to create a secret, as in
+   # Ensure to create a secret named ceph-cluster-<cluster-id>, based on
```
Fixed.

ShyamsundarR commented on examples/rbd/storageclass.yaml:
```diff
@@ -5,23 +5,14 @@ metadata:
-   # clusterID: <cluster-id>
-   # OR,
-   # if "monitors" parameter is not set, driver to get monitors from same
-   # secret as admin/user credentials. "monValueFromSecret" provides the
```
Fixed.

ShyamsundarR commented:
@dillaman I am adding the DNM flag to this. As per my understanding the following actions need to be closed before we can merge this, (reviews and acks still welcome)
- Finalize CephFS design for the same (action is on me to take this to a resolution)
- Get a reviewer who can review, with added focus, golang specifics in the change (@Madhu-1 @gman0 @humblec requesting a review with this added focus).

dillaman commented:
@ShyamsundarR Thanks!

kfox1111 commented:
We found an issue in the helm chart where it is writing configmaps to the wrong namespace.(https://github.com/ceph/ceph-csi/issues/342)

We could fix it but then would require steps to move the configmaps from the old location to the new location.

But if we have a migration plan for getting configmaps to omaps, soon, we could skip fixing the configmap issue and recommend just switching to omaps.

How soon do we thing this pr is:
 * ready for production
 * have a migration tool from configmaps to omaps

ShyamsundarR commented:
>     * Get a reviewer who can review, with added focus, golang specifics in the change (@Madhu-1 @gman0 @humblec requesting a review with this added focus).

Requesting reviews @humblec @Madhu-1 @gman0 

@phlogistonjohn any help is appreciated.

humblec commented on cmd/cephcsi.go:
```diff
@@ -38,20 +38,20 @@ const (
-	configRoot    = flag.String("configroot", "/etc/csi-config", "directory in which CSI specific Ceph"+
-		" cluster configurations are present, OR the value \"k8s_objects\" if present as kubernetes secrets")
+	instanceID    = flag.String("instanceid", "", "Unique ID distinguishing this instance of Ceph CSI among other"+
+		" instances, when sharing Ceph clusters across CSI instances for provisioning")
```
@ShyamsundarR  Is it possible or well defined solution which allows same Ceph Cluster for multple provisioners?

humblec commented on pkg/rbd/controllerserver.go:
```diff
@@ -87,10 +53,17 @@ func (cs *ControllerServer) validateVolumeReq(req *csi.CreateVolumeRequest) erro
 }
 
-func parseVolCreateRequest(req *csi.CreateVolumeRequest) (*rbdVolume, error) {
+func (cs *ControllerServer) genVolFromCreateRequest(req *csi.CreateVolumeRequest) (*rbdVolume, error) {
```
@ShyamsundarR may be I am missing something why the name `genVol..` , Isnt it actually do `parseVolCreateRequest' ?

ShyamsundarR replied:
Changed back to original.

humblec commented on pkg/rbd/controllerserver.go:
```diff
@@ -111,94 +84,80 @@ func parseVolCreateRequest(req *csi.CreateVolumeRequest) (*rbdVolume, error) {
 
-	rbdVol.VolSize = util.RoundUpToMiB(volSizeBytes)
+	// always round up the request size in bytes to the nearest MiB
+	rbdVol.VolSize = util.MiB * util.RoundUpToMiB(volSizeBytes)
```
My understanding was that the smallest unit is "Bytes" or is it "MiB"

humblec commented on pkg/rbd/controllerserver.go:
```diff
@@ -111,94 +84,80 @@ func parseVolCreateRequest(req *csi.CreateVolumeRequest) (*rbdVolume, error) {
 	if err := cs.validateVolumeReq(req); err != nil {
 		return nil, err
 	}
+
```
I dont think this new line is required here.

ShyamsundarR replied:
Removed the line space

humblec commented on pkg/rbd/controllerserver.go:
```diff
@@ -265,93 +224,75 @@ func (cs *ControllerServer) DeleteVolume(ctx context.Context, req *csi.DeleteVol
-	keys := make([]string, 0)
-	for k := range rbdVolumes {
-		keys = append(keys, k)
+		if err := unreserveVol(rbdVol, req.GetSecrets()); err != nil {
```
reserve and unreserveVol function name looks vague to me .

humblec replied:
may be some function names with `store` or `map`?

ShyamsundarR replied:
In my refactor patch this is renamed as `ReserveName` and `UndoReservation` and `CheckReservation` for reservation of a name, undoing the same, and checking reservation of a name. I an NOT posting that patch yet as it is a refactor and I do not want to derail the review progress made here, but the patchis available [here](https://github.com/ShyamsundarR/ceph-csi/commit/c987e8bc7c8999e1072db1a902e64efb256abae1).

I am assuming this renaming is acceptable.

humblec commented on pkg/rbd/controllerserver.go:
```diff
@@ -265,93 +224,75 @@ func (cs *ControllerServer) DeleteVolume(ctx context.Context, req *csi.DeleteVol
+	}
+
+	if len(req.VolumeCapabilities) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Empty volume capabilities in request")
```
s/Empty/empty

ShyamsundarR replied:
Discussion moved to [this](https://github.com/ceph/ceph-csi/pull/312#discussion_r282535932) comment for the issue pointed out.

humblec commented on pkg/rbd/controllerserver.go:
```diff
@@ -525,104 +466,50 @@ func (cs *ControllerServer) DeleteSnapshot(ctx context.Context, req *csi.DeleteS
 
+	// lock out parallel create requests against the same snap name as we
+	// cleanup the image and associated omaps for the same
+	snapshotNameMutex.LockKey(rbdSnap.RequestName)
```
Isnt this handled in ceph ?  do we need to serialize request here in CSI ?

ShyamsundarR replied:
Parallel PVC creation rates should not be affected by this additional mutex.

The lock contention issue really depends on the granularity of mutexes in use. As we use a set of mutexes and as pointed out in the issue #279 we need to increase the number of mutexes that we can use to address the initial lock contention problem.

ShyamsundarR replied:
The mutex against the RequestName, which is the SnapshotName as sent by the CO system, is required to ensure the RADOS omaps are protected against parallel updates from a creation path when being accessed by the delete path.

Creation, stores a `key` with the passed in VolumeName (or RequestName in this context) with the image UUID as its `value` (in the CSI directory omap). The next step is to update the image UUID directory (omap) to store a key with the RequestName (the back pointer). Finally the image with the said UUID is created. Hence, this is a 3 step process.

Deletion operates in the reverse, deleting the image first, then the image UUIDDirectory (omap), and finally removing the VolumeName key from the CSI directory omap.

Hence in order to protect against out of order updates to the above objects and keys, for a given RequestName, the name based lock is acquired.

This is critical and **cannot** be circumvented currently.

Performance concerns occur **only** when the mutex is contended, and that can happen if the name based hashed mutex collides (or SnapshotNames collide (VolumeNames should not as the CO system should protect against the same as per the specification)). This can be alleviated by increasing the number of mutexes as detailed in the issue #279.

Tagging @dillaman  to the comment to review reasoning above.

dillaman replied:
I guess the bigger question is whether or not it's a legal operation for the CO to issue a delete before the create has completed? RBD doesn't currently offer any protection against racing creates/deletes since we've usually had a solid orchestration system above that would prevent things like that. About the closest issue I've had to deal with was a user w/ a home-grown orchestration system that resulted in racing image deletions against the same image. 

In a perfect world, the Ceph cluster would perform the directory updates atomically by performing two omap requests in a single atomic operation against the same object (that's what librbd does internally when storing a bi-directional name<->uuid mapping in the "rbd_directory" object). 

Another option would be to use Ceph locks (rados lock get), but once you start thinking about how to recover from a dead process holding a lock, things will start to get very complicated and ugly within ceph-csi.

[1] https://github.com/ceph/go-ceph

ShyamsundarR replied:
On the legality of a Delete before Create, the spec has answers and it is logically not feasible to do so [1] as delete requires the volume ID which only a successful create can return.

But, there seems to be a possibility on a Create for a volume name that already exists [2] (but with different properties than what is being requested), which can race with a legal Delete of the existing volume. At which point the racy behavior may occur.

If the current locking granularity (as it is name hash based locks [3]) is the problem, then we may need to support granular locks based on ref counted lock objects created for each name/volID instance in flight. If that is acceptable then it can be the basis of a solution in the future?

[1] Request timeouts and negation: https://github.com/container-storage-interface/spec/blob/master/spec.md#timeouts

[2] Already existing but different properties for the existing volume: https://github.com/container-storage-interface/spec/blob/master/spec.md#createvolume-errors

[3] Kubernetes keymutex: https://github.com/kubernetes/utils/blob/master/keymutex/hashed.go

ShyamsundarR replied:
@humblec the current thought process is that we move to more granular locks as a part of fixing #279 and hence the lock contention here and in other existing places is not an bottleneck in the future. Hope it addresses your concern.

humblec replied:
@ShyamsundarR the locking concerns and the race conditions I pointed out in our discussion still exist ( thanks for documenting some of it above on behalf) , however lets handle it differently. I am taking it in my todo list. for now, lets not worry about it in this PR .

humblec commented on pkg/rbd/nodeserver.go:
```diff
@@ -47,14 +48,25 @@ type NodeServer struct {
+	if req.GetVolumeId() == "" {
+		return nil, status.Error(codes.InvalidArgument, "Empty volume ID in request")
+	}
+
```
s/Empty/s in above if's..

ShyamsundarR replied:
**NOTE**: Retaining one comment on the capitalization of first letter of messages in status.Error for ease of discussion.

As [pointed](https://github.com/ceph/ceph-csi/pull/312#discussion_r282756648) out by @Madhu-1 error strings need to start with a lower case letter in golang. This is also enforced by the linter.

These errors seem to be RPC error messages, **should they or do they follow the same convention** is something to consider before fixing it.

I assume we open a new issue against the code, to fix this across rather than address part of them here and rest later. IOW, I intend to leave added status.Errors in the current form than correcting them to adhere to the first letter being lower case.

@humblec or @Madhu-1 if you feel strongly that this should be corrected in this PR, for messages added with this PR, let me know.

humblec commented on pkg/rbd/rbd.go:
```diff
@@ -89,15 +184,24 @@ func NewNodeServer(d *csicommon.CSIDriver, containerized bool) (*NodeServer, err
-	confStore, err = util.NewConfigStore(configRoot)
-	if err != nil {
-		klog.Fatalln("Failed to initialize config store.")
+	csiConfigFile = "/etc/ceph-csi-config/config.json"
```
Make the default value a CONST.

ShyamsundarR replied:
Done.

humblec commented on pkg/rbd/rbd_util.go:
```diff
@@ -205,10 +185,491 @@ func rbdStatus(pOpts *rbdVolume, userID string, credentials map[string]string) (
+	// provide an easy way to check when this function can be reused
+	if rbdVol.RequestName == "" || rbdVol.Monitors == "" || rbdVol.AdminID == "" ||
+		rbdVol.Pool == "" || rbdVol.ClusterID == "" || rbdVol.VolSize == 0 {
+		return false, errors.New("missing information in rbdVolume to check for existence")
```
This if condition looks bit ugly to me.

ShyamsundarR replied:
Checking each element and printing what is missing can be tedious, this is a developer log when the function is reused. Hence, printed the entire structure contents on errors, to help aid debugging instead.

ShyamsundarR replied:
Addressed now with a validate routine for rbdSnap and rbdVol to throw exact field that is missing and handle the error.

humblec commented on pkg/rbd/rbd_util.go:
```diff
@@ -467,141 +1034,252 @@ func protectSnapshot(pOpts *rbdSnapshot, adminID string, credentials map[string]
 	if err != nil {
-		return err
+		klog.Errorf("failed getting information for image (%s): (%s)", poolName+"/"+imageName, err)
+		if strings.Contains(string(stdout), "rbd: error opening image "+imageName+
```
Why we need this check here ?

ShyamsundarR replied:
The check is to detect missing images, and segregate them from other classes of errors. We need this to return success for successive DeleteVolume requests in case the first one times out, but the work is actually done. If we cannot detect image missing errors it would appear to the CO system as internal errors or other errors.

The above is one case of the same, which is needed in the code. The error itself can be used generically to detect that such an image does not exist.

humblec commented on pkg/util/csiconfig.go:
```diff
@@ -0,0 +1,97 @@
+							return "", fmt.Errorf("malformed clusterID section in csi"+
+								" configuration (%v)", cvalue)
+						}
+					}
```
This block has to be simplified, series of `for, if, switch` does not looks good from coding pov.

ShyamsundarR replied:
This block should at least have had better comments, anyway here is the problem and possible solutions,

The JSON structure that I have chosen has variable key names, as the key names are the `<clusterID>`, which makes it difficult to parse into a strongly typed structure in go, and hence has to be parsed as an interface and then type checked and parsed further, leading to the loops above.

The solution to simplifying the code here would be as follows,
- Write a series of nested functions, each dealing with a part of the type casting and passing it further
OR,
- Is there a way on go to parse such structures more elegantly?
OR,
- Change the config JSON to the following known key values, and use a strongly typed JSON structure to parse the same,
```
[
    {
      "clusterID": "<clusterID>",
      "monitors": [
          "<MONValue1>",
          "<MONValue2>",
          ...
          "<MONValueN>"
      ]
    },
    ...
]
```

I prefer changing the JSON to a strongly types structure, keeps it simple for other config managers to parse and update as well.

Let me know if you have a preference or another solution for the problem.

ShyamsundarR replied:
Changed configuration to reflect new JSON structure instead, thereby simplifying the code here.

humblec commented:
@ShyamsundarR Started the review, there is also conflict. Can you correct the conflict and update the PR. Btw, I see there are good number of changes are which are just intendation correction. Can you remove those changes from this PR for better review look?

ShyamsundarR commented:
> @ShyamsundarR Started the review, there is also conflict. Can you correct the conflict and update the PR. Btw, I see there are good number of changes which are just indentation correction. Can you remove those changes from this PR for better review look?

@humblec I would be updating the PR addressing any review comments and squashing the commits into one, at which point I intend to address the conflicts as well and also would rebase to tip of master. Is this fine? Or, do you want me to rebase right away?

I went through the PR attempting to find the indentation corrections only, and found about 18 lines that are such changes (out of ~2200 additions in all). Some of these 18 are in YAML where there were trailing white spaces, some for lines that were too long, and a few to introduce a line break at some logical point in the code.

Further, I used the diff settings in github to ignore whitespaces and counted line deltas, other than 5 files no other files contain whitespace only changes when comparing the changed lines without the said diff settings.

Some structures and assignment blocks are auto indented when adding or removing a variable whose length is different, these white space changes are out of my control (unless I use a plain text editor to correct them back to their original settings).

I do understand the problem white space only changes can cause in reviews, but the quantum in this PR seems quite low (~18 as I saw it).

Do you really want me to remove these in the PR? As this calls for very careful looking and removing just these changes. If they are hindering review, could you change the diff settings for this PR to [ignore white spaces](https://github.blog/2018-05-01-ignore-white-space-in-code-review/) and take a look?

Madhu-1 commented on pkg/rbd/rbd_util.go:
```diff
@@ -467,141 +1034,252 @@ func protectSnapshot(pOpts *rbdSnapshot, adminID string, credentials map[string]
+
+	return snpInfo, ErrSnapNotFound{snapName, fmt.Errorf("snap (%s) for image (%s) not found",
+		snapName, poolName+"/"+imageName)}
 }
```
This file looks huge to me, can we split it up if possible?

ShyamsundarR replied:
Split out the functions that deal with keys and objects, IOW maintain the journal. Should now be reasonable in size and logically split.

FYI, further as the CephFS implementation is done, the voljournal should move to util instead of being retained in the rbd directory.

Madhu-1 commented on pkg/util/cephcmds.go:
```diff
@@ -0,0 +1,288 @@
+}
+
+// ErrObjectExists is returned when named omap is already present in rados
+type ErrObjectExists struct {
```
as said above move this error structs  to new file

ShyamsundarR replied:
Done

ShyamsundarR commented on deploy/rbd/helm/templates/nodeplugin-daemonset.yaml:
```diff
@@ -41,7 +41,7 @@ spec:
                   command: [
                     "/bin/sh", "-c",
-                    'rm -rf /registration/{{ .Values.driverName }} 
+                    'rm -rf /registration/{{ .Values.driverName }}
```
Done

ShyamsundarR commented on docs/deploy-rbd.md:
````diff
@@ -101,6 +88,18 @@ Those manifests deploy service accounts, cluster roles and cluster role
 
+**Deploy ConfigMap for CSI plugins:**
+
+```bash
````
Retained `bash` as all other snippets use `bash` in this file and did not want to change more contents in this PR.

ShyamsundarR commented on pkg/cephfs/volumemounter.go:
```diff
@@ -25,6 +25,7 @@ import (
 	"sync"
 
+	"github.com/ceph/ceph-csi/pkg/util"
 	"k8s.io/klog"
```
Done

ShyamsundarR commented on pkg/rbd/controllerserver.go:
```diff
@@ -111,94 +84,80 @@ func parseVolCreateRequest(req *csi.CreateVolumeRequest) (*rbdVolume, error) {
+		if err != nil {
+			errDefer := unreserveVol(rbdVol, req.GetSecrets())
+			if errDefer != nil {
+				klog.Warningf("failed undoing reservation of volume: %s", req.GetName())
```
Done

ShyamsundarR commented on pkg/rbd/controllerserver.go:
```diff
@@ -244,15 +200,18 @@ func (cs *ControllerServer) checkSnapshot(req *csi.CreateVolumeRequest, rbdVol *
 		return status.Error(codes.Internal, err.Error())
 	}
-	klog.V(4).Infof("create volume %s from snapshot %s", req.GetName(), rbdSnap.SnapName)
+	klog.V(4).Infof("create volume %s from snapshot %s", req.GetName(), rbdSnap.RbdSnapName)
```
While coding and even debugging various failures, as we carry multiple names in this structure (request name image name etc. and also in the snaps structure), it was very useful to add that suffix to this element to denote that this is the name on rbd as opposed to the vol name or other names.

Hence, I would like to retain the new name, even if it carries the same suffix as the structure. If you have a strong preference, let me know and I will change it across the code base appropriately.

ShyamsundarR commented on pkg/rbd/controllerserver.go:
```diff
@@ -265,93 +224,75 @@ func (cs *ControllerServer) DeleteVolume(ctx context.Context, req *csi.DeleteVol
+	klog.V(4).Infof("deleting image %s", rbdVol.RbdImageName)
+	if err := deleteImage(rbdVol, rbdVol.AdminID, req.GetSecrets()); err != nil {
+		// TODO: can we detect "already deleted" situations here and proceed?
+		klog.V(3).Infof("failed to delete rbd image: %s/%s with error: %v",
```
Done

ShyamsundarR commented on pkg/rbd/controllerserver.go:
```diff
@@ -368,100 +309,93 @@ func (cs *ControllerServer) ValidateVolumeCapabilities(ctx context.Context, req
-					ReadyToUse: true,
-				},
-			}, nil
+	// Fetch source volume information
```
Done

ShyamsundarR commented on pkg/rbd/controllerserver.go:
```diff
@@ -475,41 +409,48 @@ func (cs *ControllerServer) validateSnapshotReq(req *csi.CreateSnapshotRequest)
-			} else {
-				klog.Warningf("failed to create snapshot: %v", err)
-				return err
+		klog.Warningf("failed to create snapshot: %v", err)
```
Done.

ShyamsundarR commented on pkg/rbd/controllerserver.go:
```diff
@@ -475,41 +409,48 @@ func (cs *ControllerServer) validateSnapshotReq(req *csi.CreateSnapshotRequest)
 
+	err = protectSnapshot(rbdSnap, rbdSnap.AdminID, secret)
+	if err != nil {
+		return errors.New("snapshot created but failed to protect snapshot")
```
Done

ShyamsundarR commented on pkg/rbd/controllerserver.go:
```diff
@@ -475,41 +409,48 @@ func (cs *ControllerServer) validateSnapshotReq(req *csi.CreateSnapshotRequest)
+	}()
+
+	err = getSnapshotMetadata(rbdSnap, rbdSnap.AdminID, secret)
+	if err != nil {
```
In this instance, as this is a cs function, changed the error to return a proper gRPC error code, and hence retained the last err block.

ShyamsundarR commented on pkg/rbd/nodeserver.go:
```diff
@@ -103,22 +115,15 @@ func (ns *NodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublis
-		volName = s[len(s)-1]
+	var vi util.CsiIdentifier
+
+	err := vi.DecomposeCsiID(req.GetVolumeId())
```
Done (across the code base for vars and function names that had `Csi` to `CSI`

ShyamsundarR commented on pkg/rbd/rbd.go:
```diff
@@ -47,8 +135,16 @@ type Driver struct {
+	csiConfigFile string
+	// CsiInstanceID is the instance ID that is unique to an instance of CSI, used when sharing
+	// ceph clusters across CSI instances, to differentiate omap names per CSI instance
+	CsiInstanceID = "default"
```
Done (across the code base for vars and function names that had `Csi` to `CSI`

ShyamsundarR commented on pkg/rbd/rbd.go:
```diff
@@ -89,15 +184,24 @@ func NewNodeServer(d *csicommon.CSIDriver, containerized bool) (*NodeServer, err
+
+	// Create ceph.conf for use with CLI commands
+	if err = util.WriteCephConfig(); err != nil {
+		klog.Fatalf("failed to write ceph configuration file")
```
Done

ShyamsundarR commented on pkg/rbd/rbd_util.go:
```diff
@@ -205,10 +185,491 @@ func rbdStatus(pOpts *rbdVolume, userID string, credentials map[string]string) (
+	// provide an easy way to check when this function can be reused
+	if rbdSnap.RequestName == "" || rbdSnap.Monitors == "" || rbdSnap.AdminID == "" ||
+		rbdSnap.Pool == "" || rbdSnap.RbdImageName == "" || rbdSnap.ClusterID == "" {
+		return false, errors.New("missing information in rbdSnapshot to check for existence")
```
Done. Logged the entire structure instead of each element, to reduce code checks and reporting for each element.

ShyamsundarR commented on pkg/rbd/rbd_util.go:
```diff
@@ -205,10 +185,491 @@ func rbdStatus(pOpts *rbdVolume, userID string, credentials map[string]string) (
+
+// ErrVolNameConflict is generated when a requested CSI volume name already exists on RBD but with
+// different properties, and hence is in conflict with the passed in CSI volume name
+type ErrVolNameConflict struct {
```
Done

ShyamsundarR commented on pkg/rbd/rbd_util.go:
```diff
@@ -205,10 +185,491 @@ func rbdStatus(pOpts *rbdVolume, userID string, credentials map[string]string) (
+	// Fetch on-disk image attributes and compare against request
+	err = updateVolWithImageInfo(rbdVol, credentials)
+	if err != nil {
+		if _, ok := err.(ErrImageNotFound); ok {
```
I thought about this based on the comment provided, was not able to come up with a satisfactory snippet that can be repeated.

Further the functions that are dealing with the keys and objects is split into its own file. When reusing this for FS I see opportunities to reduce repetitions (including functions like reserveVol/Snap etc.), and will tackle this with the FS change.

Let me know if this is acceptable.

ShyamsundarR commented on pkg/rbd/rbd_util.go:
```diff
@@ -205,10 +185,491 @@ func rbdStatus(pOpts *rbdVolume, userID string, credentials map[string]string) (
+	err = util.RemoveObject(rbdSnap.Monitors, rbdSnap.AdminID, key, rbdSnap.Pool, rbdSnapOMapPrefix+snapUUID)
+	if err != nil {
+		if _, ok := err.(util.ErrObjectNotFound); !ok {
+			klog.V(4).Infof("failed removing oMap %s (%s)", rbdSnapOMapPrefix+snapUUID, err)
```
Done

ShyamsundarR commented on pkg/rbd/rbd_util.go:
```diff
@@ -205,10 +185,491 @@ func rbdStatus(pOpts *rbdVolume, userID string, credentials map[string]string) (
+	err = util.RemoveOMapKey(rbdSnap.Monitors, rbdSnap.AdminID, key, rbdSnap.Pool,
+		csiSnapsDirectory, csiSnapNameKeyPrefix+rbdSnap.RequestName)
+	if err != nil {
+		klog.V(4).Infof("failed removing oMap key %s (%s)", csiSnapNameKeyPrefix+rbdSnap.RequestName, err)
```
Done

ShyamsundarR commented on pkg/rbd/rbd_util.go:
```diff
@@ -205,10 +185,491 @@ func rbdStatus(pOpts *rbdVolume, userID string, credentials map[string]string) (
+	err = util.RemoveObject(rbdVol.Monitors, rbdVol.AdminID, key, rbdVol.Pool, rbdImageOMapPrefix+imageUUID)
+	if err != nil {
+		if _, ok := err.(util.ErrObjectNotFound); !ok {
+			klog.V(4).Infof("failed removing oMap %s (%s)", rbdImageOMapPrefix+imageUUID, err)
```
Done

ShyamsundarR commented on pkg/rbd/rbd_util.go:
```diff
@@ -205,10 +185,491 @@ func rbdStatus(pOpts *rbdVolume, userID string, credentials map[string]string) (
+	err = util.RemoveOMapKey(rbdVol.Monitors, rbdVol.AdminID, key, rbdVol.Pool,
+		csiVolsDirectory, csiVolNameKeyPrefix+rbdVol.RequestName)
+	if err != nil {
+		klog.V(4).Infof("failed removing oMap key %s (%s)", csiVolNameKeyPrefix+rbdVol.RequestName, err)
```
Done

ShyamsundarR commented on pkg/rbd/rbd_util.go:
```diff
@@ -205,10 +185,491 @@ func rbdStatus(pOpts *rbdVolume, userID string, credentials map[string]string) (
+func reserveOMapName(monitors, adminID, key, poolName, oMapNamePrefix string) (string, error) {
+	var iterUUID string
+
+	maxAttempts := 5
```
No reason, if there is a UUID clash that is not resolved in 5 retries, it may never be :) . Further, the CO would retry anyway to get another 5 tries to get a new UUID.

ShyamsundarR commented on pkg/rbd/rbd_util.go:
```diff
@@ -205,10 +185,491 @@ func rbdStatus(pOpts *rbdVolume, userID string, credentials map[string]string) (
+	poolID, err := util.GetPoolID(rbdSnap.Monitors, rbdSnap.AdminID, key,
+		rbdSnap.Pool)
+	if err != nil {
+		klog.Errorf("Error fetching pool ID (%s)", err)
```
I was logging for all `GetPoolID` call failures, but as the function itself logs, removed the same across other calls as well.

ShyamsundarR commented on pkg/rbd/rbd_util.go:
```diff
@@ -205,10 +185,491 @@ func rbdStatus(pOpts *rbdVolume, userID string, credentials map[string]string) (
+			klog.Warningf("reservation failed for volume: %s", rbdSnap.RequestName)
+			errDefer := unreserveSnap(rbdSnap, credentials)
+			if errDefer != nil {
+				klog.Warningf("failed undoing reservation of snapshot: %s", rbdSnap.RequestName)
```
Done

ShyamsundarR commented on pkg/rbd/rbd_util.go:
```diff
@@ -205,10 +185,491 @@ func rbdStatus(pOpts *rbdVolume, userID string, credentials map[string]string) (
+			klog.Warningf("reservation failed for volume: %s", rbdVol.RequestName)
+			errDefer := unreserveVol(rbdVol, credentials)
+			if errDefer != nil {
+				klog.Warningf("failed undoing reservation of volume: %s", rbdVol.RequestName)
```
Done

ShyamsundarR commented on pkg/rbd/rbd_util.go:
```diff
@@ -217,86 +678,248 @@ func deleteRBDImage(pOpts *rbdVolume, adminID string, credentials map[string]str
+		"--key=" + key}
+	output, err = execCommand("rbd", args)
 	if err != nil {
+		klog.Errorf("failed to delete rbd image: %v, command output: %s", err, string(output))
```
Afaics, this one does not. Also this was used earlier as is (IOW, output logging is not added in this PR).

It is a valid concern though, should we open another issue to do the audit across the code and handle any exceptions through that?

ShyamsundarR replied:
Added an issue #376

ShyamsundarR commented on pkg/rbd/rbd_util.go:
```diff
@@ -217,86 +678,248 @@ func deleteRBDImage(pOpts *rbdVolume, adminID string, credentials map[string]str
+	}
+
+	rbdSnap.CreatedAt, err = ptypes.TimestampProto(tm)
+	if err != nil {
```
Done

ShyamsundarR commented on pkg/rbd/rbd_util.go:
```diff
@@ -217,86 +678,248 @@ func deleteRBDImage(pOpts *rbdVolume, adminID string, credentials map[string]str
+		return err
+	}
+
+	switch imageInfo.Format {
```
Done

ShyamsundarR commented on pkg/rbd/rbd_util.go:
```diff
@@ -217,86 +678,248 @@ func deleteRBDImage(pOpts *rbdVolume, adminID string, credentials map[string]str
+
+	err := vi.DecomposeCsiID(rbdSnap.SnapID)
+	if err != nil {
+		klog.V(4).Infof("error decoding snapshot ID (%s) (%s)", err, rbdSnap.SnapID)
```
Done

ShyamsundarR commented on pkg/rbd/rbd_util.go:
```diff
@@ -217,86 +678,248 @@ func deleteRBDImage(pOpts *rbdVolume, adminID string, credentials map[string]str
+	}
+
+	err = updateSnapWithImageInfo(rbdSnap, credentials)
+	if err != nil {
```
Done

ShyamsundarR commented on pkg/rbd/rbd_util.go:
```diff
@@ -217,86 +678,248 @@ func deleteRBDImage(pOpts *rbdVolume, adminID string, credentials map[string]str
+	}
+
+	err = updateVolWithImageInfo(rbdVol, credentials)
+	if err != nil {
```
Done

ShyamsundarR commented on pkg/rbd/rbd_util.go:
```diff
@@ -205,10 +185,491 @@ func rbdStatus(pOpts *rbdVolume, userID string, credentials map[string]string) (
+	snapUUID, err := util.GetOMapValue(rbdSnap.Monitors, rbdSnap.AdminID,
+		key, rbdSnap.Pool, csiSnapsDirectory, csiSnapNameKeyPrefix+rbdSnap.RequestName)
+	if err != nil {
+		return false, nil
```
The return of `nil` error is intentional as there is no error if image is not present. Hence the bool is `false`.

However, when reviewing this comment, found that we need to specifically detect and return a success only if image is not found, not when any other error is encountered, as that may cause overwrites of legal objects and keys. Fixed this by returning the said `false, nil` only when the error is specifically missing key.

ShyamsundarR commented on pkg/rbd/rbd_util.go:
```diff
@@ -205,10 +185,491 @@ func rbdStatus(pOpts *rbdVolume, userID string, credentials map[string]string) (
+	imageUUID, err := util.GetOMapValue(rbdVol.Monitors, rbdVol.AdminID,
+		key, rbdVol.Pool, csiVolsDirectory, csiVolNameKeyPrefix+rbdVol.RequestName)
+	if err != nil {
+		return false, nil
```
The return of `nil` error is intentional as there is no error if image is not present. Hence the bool is `false`.

However, when reviewing this comment, found that we need to specifically detect and return a success only if image is not found, not when any other error is encountered, as that may cause overwrites of legal objects and keys. Fixed this by returning the said `false, nil` only when the error is specifically missing key.

ShyamsundarR commented:
@Madhu-1 @humblec completed addressing review comments provided till now. 

Added some comments on reviews not addressed for discussion.

Requesting further reviews and discussion to take this to completion. Thanks!

humblec commented on pkg/rbd/controllerserver.go:
```diff
@@ -87,10 +52,17 @@ func (cs *ControllerServer) validateVolumeReq(req *csi.CreateVolumeRequest) erro
 	}
+	options := req.GetParameters()
+	if value, ok := options["clusterID"]; !ok || len(value) == 0 {
+		return status.Error(codes.InvalidArgument, "Missing or empty cluster ID to provision volume from")
```
s/Missing/missing

ShyamsundarR replied:
I thought we are discussing this [here](https://github.com/ceph/ceph-csi/pull/312#discussion_r282535932), IOW, do you want me to change the strings that I added in this patch to capitalized first letter OR are we dealing with this in a separate patch as the code base needs a revision on the same? Please let me know so that I can handle it appropriately.

humblec commented on pkg/rbd/controllerserver.go:
```diff
@@ -265,93 +223,75 @@ func (cs *ControllerServer) DeleteVolume(ctx context.Context, req *csi.DeleteVol
-		if err, ok := err.(*util.CacheEntryNotFound); ok {
-			klog.V(3).Infof("metadata for volume %s not found, assuming the volume to be already deleted (%v)", volumeID, err)
+	if err := genVolFromVolID(rbdVol, volumeID, req.GetSecrets()); err != nil {
+		// If image key is missing, there can be no unreserve as request name is unknown, and also
```
some how failing to understand `no unreserve..etc` , may be reword it ?

ShyamsundarR replied:
Updated comments in the area appropriately.

humblec commented on pkg/rbd/controllerserver.go:
```diff
@@ -265,93 +223,75 @@ func (cs *ControllerServer) DeleteVolume(ctx context.Context, req *csi.DeleteVol
+	// Deleting rbd image
+	klog.V(4).Infof("deleting image %s", rbdVol.RbdImageName)
+	if err := deleteImage(rbdVol, rbdVol.AdminID, req.GetSecrets()); err != nil {
+		// TODO: can we detect "already deleted" situations here and proceed?
```
If we dont see the image , thats a success.

ShyamsundarR replied:
This is an [existing](https://github.com/ceph/ceph-csi/blob/f60a07ae8268136f40de3d62537f4f8b020f0ec4/pkg/rbd/controllerserver.go#L290) TODO, I am not touching this TODO yet, and do not intend to do so. Do you want me to fix this as a part of this patch?

ShyamsundarR replied:
Removed the TODO, already addressed in code above, hence older TODO is no longer valid.

humblec commented on pkg/rbd/controllerserver.go:
```diff
@@ -368,100 +308,92 @@ func (cs *ControllerServer) ValidateVolumeCapabilities(ctx context.Context, req
-			}, nil
+	// Fetch source volume information
+	//var rbdVol *rbdVolume
+	rbdVol := new(rbdVolume)
```
I think line 323 exist as part of memory allocation experiment :), please remove it.

ShyamsundarR replied:
Removed.

humblec commented on pkg/rbd/controllerserver.go:
```diff
@@ -368,100 +308,92 @@ func (cs *ControllerServer) ValidateVolumeCapabilities(ctx context.Context, req
 	}
+	defer func() {
+		if err != nil {
+			klog.Warningf("snapshot failed, undoing reservation of snap name: %s", req.GetName())
```
Why warning ?

ShyamsundarR replied:
Would you prefer Info instead? This path should be arrived at post reservation in case of errors, the original errors are not logged here, but a undo of the reservation is being logged.

ShyamsundarR replied:
Removed, not a pattern elsewhere to log, so not adding it here.

humblec commented on pkg/rbd/errors.go:
```diff
@@ -0,0 +1,60 @@
+	err       error
+}
+
+func (e ErrImageNotFound) Error() string {
```
Hmm.. this does not look like a common pattern for describing errors and method on top.

ShyamsundarR replied:
Can you point to a common pattern to adopt? I do not know of any hence asking.

Madhu-1 replied:
@ShyamsundarR can we move this file outside of rbd, s that we can use same for cephfs also.
is this possible?

ShyamsundarR replied:
Some errors move out to util when they are shared by CephFS (in the [refactor]https://github.com/ShyamsundarR/ceph-csi/commit/c987e8bc7c8999e1072db1a902e64efb256abae1) patch) some stay (e.g ErrImageNotFound) as they are RBD specific. So in this PR I cannot move it to a common file.

BTW, there is another common [file](https://github.com/ceph/ceph-csi/pull/312/files#diff-14a2aec463bbb3baabb04222dabe7fc7) in util with other errors already.

ShyamsundarR replied:
@Madhu-1 pointed out to this source for error definition https://golang.org/src/errors/errors.go and as per this, the way errors are defined looks right. @humblec can you clarify if this looks right?

Madhu-1 commented on docs/deploy-rbd.md:
```diff
@@ -32,52 +32,39 @@ Option | Default value | Description
-
 ## Deployment with Kubernetes
 
 Requires Kubernetes 1.11
```
This need to be kube >=1.13

ShyamsundarR replied:
This is not updated as a part of this patch, I am testing against v1.13 and also do not have any specific Kube version dependent changes to call out a different version.

Do you want me to change this as a part of this PR, or should it be a separate edit?

Madhu-1 replied:
Agreed, This can a separate patch or it can be updated as part of this patch

ShyamsundarR commented on docs/deploy-rbd.md:
```diff
@@ -32,52 +32,39 @@ Option | Default value | Description
-`monitors` | one of `monitors`, `clusterID` or `monValueFromSecret` must be set | Comma separated list of Ceph monitors (e.g. `192.168.100.1:6789,192.168.100.2:6789,192.168.100.3:6789`)
-`monValueFromSecret` | one of `monitors`, `clusterID` or and `monValueFromSecret` must be set | a string pointing the key in the credential secret, whose value is the mon. This is used for the case when the monitors' IP or hostnames are changed, the secret can be updated to pick up the new monitors.
-`clusterID` | one of `monitors`, `clusterID` or `monValueFromSecret` must be set | String representing a Ceph cluster, must be unique across all Ceph clusters in use for provisioning, cannot be greater than 36 bytes in length, and should remain immutable for the lifetime of the Ceph cluster in use
+`clusterID` | yes | String representing a Ceph cluster, must be unique across all Ceph clusters in use for provisioning, cannot be greater than 36 bytes in length, and should remain immutable for the lifetime of the Ceph cluster in use
```
Almost all first letters of the description column in the deploy-rbd.md doc is capitalized, as seen [here](https://github.com/ceph/ceph-csi/blob/master/docs/deploy-rbd.md#configuration), are you suggesting the first "String" word be in lower case as in "string"

Madhu-1 commented:
@humblec PTAL.

kfox1111 commented:
Is there a migration path defined yet?

ShyamsundarR commented:
> Is there a migration path defined yet?

I want to address this in a few ways, hence opened an issue #378 that talks about the same with the label that it is part of the release 1.1.

humblec commented:
@ShyamsundarR To summarize my thoughts, we need to do some more code improvement here and I would like to sort out the locking stuff in another PR. I am making a design for the same. However these can be  follow up PRs and lets unblock this for now. So, I am fine with addressing rest in subsequent work. In short, approved.
